titus-common/src/main/java/com/netflix/titus/common/framework/simplereconciler/internal/EventDistributor.java

```diff
@@ -0,0 +1,270 @@
+                cancelled.add(id);
+            }
+        });
+        activeSinks.keySet().removeAll(cancelled);
+    }
+
+    private List<SimpleReconcilerEvent<DATA>> mergeSnapshotAndPendingEvents(List<SimpleReconcilerEvent<DATA>> snapshot,
+                                                                            List<SimpleReconcilerEvent<DATA>> events) {
+        if (events.isEmpty()) {
+            return snapshot;
+        }
+        Map<String, SimpleReconcilerEvent<DATA>> eventsById = new HashMap<>();
+        events.forEach(event -> eventsById.put(event.getId(), event));
+        List<SimpleReconcilerEvent<DATA>> mergeResult = new ArrayList<>();
+        for (SimpleReconcilerEvent<DATA> event : snapshot) {
+            SimpleReconcilerEvent<DATA> fromQueue = eventsById.get(event.getId());
+            if (fromQueue == null || fromQueue.getTransactionId() < event.getTransactionId()) {
+                mergeResult.add(event);
+            } else if (fromQueue.getKind() != SimpleReconcilerEvent.Kind.Removed) {
+                mergeResult.add(new SimpleReconcilerEvent<>(
+                        SimpleReconcilerEvent.Kind.Added,
+                        fromQueue.getId(),
+                        fromQueue.getData(),
+                        fromQueue.getTransactionId()
+                ));
+            }
+        }
+        return mergeResult;
+    }
+
+    private void processEvents(List<SimpleReconcilerEvent<DATA>> events) {
+        events.forEach(event -> {
+            Set<String> cancelled = new HashSet<>();
+            activeSinks.forEach((id, holder) -> {
+                if (!holder.onNext(events)) {
+                    cancelled.add(id);
+                }
+            });
+            activeSinks.keySet().removeAll(cancelled);
+        });
+    }
+
+    private void completeEmitters() {
+        removeUnsubscribedSinks();
+        Throwable error = new IllegalStateException("Reconciler framework stream closed");
+        activeSinks.forEach((id, holder) -> holder.onError(error));
+        activeSinks.clear();
+        for (EmitterHolder next; (next = sinkQueue.poll()) != null; ) {
+            next.onError(error);
+        }
+    }
+
+    private class EmitterHolder {
+
+        private final String id;
+        private final FluxSink<List<SimpleReconcilerEvent<DATA>>> emitter;
+        private volatile boolean cancelled;
+
+        private EmitterHolder(String id, FluxSink<List<SimpleReconcilerEvent<DATA>>> emitter) {
+            this.id = id;
+            this.emitter = emitter;
+            emitter.onCancel(() -> cancelled = true);
+        }
+
+        private String getId() {
+            return id;
+        }
+
+        private boolean isCancelled() {
+            return cancelled;
+        }
+
+        private boolean onNext(List<SimpleReconcilerEvent<DATA>> event) {
+            if (cancelled) {
+                return false;
+            }
+
+            try {
+                emitter.next(event);
```

Comment:
If one of the `activeSink`/subscribers is slow would we expect the `emitter.next(event)` to be impacted? I suppose at least the reconciler `doProcess()` loop is decoupled from this so won't be impacted.

Reply:
Yes, slow consumers will have an impact here. That is why all slow consumers should subscribe on their own thread.